### PR TITLE
Add 2.1 Preflists and Write-Once Bucket Types

### PR DESCRIPTION
**Pull request by javajolt**
These features added in Riak 2.1 have been missing from the Python client up until now.

### CODE REVIEW

**Comment by javajolt** [comment]
FYI - Need to update the `riak_pb` distribution before this will work

**Comment by javajolt** [comment]
Determined that this will fail on 2.1.1 because it is missing @lukebakken's fix for preflist permissions: https://github.com/basho/riak_kv/pull/1116

File: riak/transports/feature_detect.py [hunk 1]
```diff
@@ -192,6 +194,22 @@ def datatypes(self):
         """
         return self.server_version >= versions[2.0]
 
+    def preflists(self):
+        """
+        Whether bucket/key preflists are supported.
+
+        :rtype: bool
+        """
+        return self.server_version >= versions[2.1]
```
Comment (lukebakken) [inline riak/transports/feature_detect.py]: Should this also check for security and, if enabled, require a version greater than 2.1.1 ?

Comment (javajolt) [inline riak/transports/feature_detect.py]: That's a great idea.  I just tried this with 2.1.2rc3 but it seems to report back a version of 2.1.1.  It occurred to me that buildbot also is trying to apply the permission in the configuration step, but that is failing.  So putting a check here won't help that.

File: riak/transports/pbc/connection.py [hunk 2]
```diff
@@ -165,7 +165,7 @@ def _ssl_handshake(self):
 
                     return True
                 except ssl.SSLError as e:
-                    raise SecurityError(e.library + ": " + e.reason)
```
Comment (lukebakken) [inline riak/transports/pbc/connection.py]: `library` was not an attribute of `SSLError` in my testing. Just wrapping the exception provides more information to the user.

**Comment by lukebakken** [comment]
@javajolt I made some changes to address the cipher issue.

File: riak/transports/pbc/connection.py (hunk 2 above)
Comment (javajolt) [inline riak/transports/pbc/connection.py]: Which version to Python were you using? Pretty sure that is from PyOpenSSL which is only used on 2.6 and 2.7 before 2.7.10.

Comment (lukebakken) [inline riak/transports/pbc/connection.py]:
```sh
$ python2 --version
Python 2.7.10
```

I got some sort of exception saying that `library` was not an attribute of `SSLError`. So, I just changed it to pass the exception itself to `SecurityError` and I got a lot more helpful information in the output. I will reproduce and will show the difference in just a sec.

Comment (javajolt) [inline riak/transports/pbc/connection.py]: If you could try it on 2.7.9, say, that would be helpful.  Then we'd know it works everywhere.  Have you grepped for `.library`? We should change it everywhere to be consistent.

Comment (lukebakken) [inline riak/transports/pbc/connection.py]: This is the output with python 2.7.9 and the change to pass the exception directly to the `SecurityError` ctor:

```python
  File "/home/lbakken/Projects/basho/riak-python-client/riak/transports/pbc/connection.py", line 84, in _init_security
    self._ssl_handshake()
  File "/home/lbakken/Projects/basho/riak-python-client/riak/transports/pbc/connection.py", line 168, in _ssl_handshake
    raise SecurityError(e)
SecurityError: SSLEOFError(8, u'EOF occurred in violation of protocol (_ssl.c:581)')
```

This is what happens with the original code:

```python
  File "/home/lbakken/Projects/basho/riak-python-client/riak/transports/pbc/connection.py", line 84, in _init_security
    self._ssl_handshake()
  File "/home/lbakken/Projects/basho/riak-python-client/riak/transports/pbc/connection.py", line 169, in _ssl_handshake
    raise SecurityError(e.library + ": " + e.reason)
TypeError: unsupported operand type(s) for +: 'NoneType' and 'str'
```

Comment (lukebakken) [inline riak/transports/pbc/connection.py]: I grepped for `\.library` and it didn't show up anywhere.

Comment (javajolt) [inline riak/transports/pbc/connection.py]: Note to self: Always look at the code first.  Looks like 2.7.8 actually was the last version supporting PyOpenSSL.  Which is interesting.  I should tweak the versions in tox, then, too to reflect this.

**Comment by lukebakken** [comment]
I say :shipit:

**Comment by javajolt** [comment]
+1 bdbc35f

**Comment by javajolt** [comment]
@borshop merge

**Pull request merged**